File: December/Day6/odd_even_linked_list.py
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        odd = head
        even = head.next
        temp = even
        while even and even.next:
            odd.next = even.next
            odd = odd.next
            even.next = odd.next
            even = even.next
        odd.next = temp
        return head

File: December/Day6/odd-even-linked-list.cpp
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        // If there is no node inside the list, return NULL.
        if(head == NULL) {
            return head;
        }

        // Keep odd pointer on the 1st node and even pointer on the 2nd node.
        ListNode *odd = head, *even = head->next;
        // 2nd node will be the 1st even node in the list. Take this node as head of even list.
        ListNode* evenHead = head->next;

        // At every iteration, connect odd node to next odd node and even node to next even node.
        while(even && even->next) {
            odd->next = even->next;
            even->next = even->next->next;

            // Update odd and even pointer for next iteration.
            odd = odd->next;
            even = even->next;
        }

        // Connect the odd nodes linked list to even nodes linked list.
        odd->next = evenHead;
        return head;
    }
};